### PR TITLE
New: Add no-useless-constructor rule (fixes #4785)

### LINKED ISSUE #4785: Proposal: no-useless-constructor
# Rule Proposal:  Empty or Useless ES6 Constructor

## When the rules will warn. Include a description as well as sample code.
    constructor (args...) {
      super(args...);
    }

or

    constructor () {
    }

In ES6 classes both of the above forms are redundant, [as they are the default constructor if none is specified](http://www.2ality.com/2015/02/es6-classes-final.html).  As such they can be considered Dead Code as they aren't doing anything.

Many times if code is here it is because implied constructors are not known well by the developer or they are left-overs after some other relevant non-redundant code has been removed.

This proposal is to provide and eslint rule to warn against this type of redundant dead code so it can be removed.  This rule is similar to empty method, unused method types of eslint rules.  

## Whether the rule prevents an error or is stylistic.
It is mostly stylistic, but it could prevent an error if the developer feels that this code is doing something different than the default constructor.

## Why the rule should be in the core instead of creating a custom rule.
This seems pretty core to removing dead or redundant code.

## Are you willing to create the rule yourself?
Not at this time no.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29322459-proposal-no-useless-constructor?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by alberto**

### CODE REVIEW

File: docs/rules/no-useless-constructor.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+## Rule Details
+
+This proposal is to provide a rule to warn against this type of redundant dead code so it can be removed. This rule is similar to `no-empty`.
+
+The following patterns are considered problems:
+
+```js
+/*eslint no-useless-constructor: 2*/
+/*eslint-env es6*/
```
Comment (platinumazure) [inline docs/rules/no-useless-constructor.md]: Does this still apply with ecmaVersion?

**Comment by gyandeeps** [comment]
can you add this also (under `valid` section):

```js
class A extends B {
    constructor(foo, bar){ 
         super(foo, bar, 1); 
     } 
}
```

File: docs/rules/no-useless-constructor.md (hunk 1 above)
Comment (alberto) [inline docs/rules/no-useless-constructor.md]: I don't know what you mean.

**Comment by alberto** [comment]
@gyandeeps test added

File: docs/rules/no-useless-constructor.md (hunk 1 above)
Comment (nzakas) [inline docs/rules/no-useless-constructor.md]: @platinumazure yes, this still works. 

@alberto http://eslint.org/docs/2.0.0/user-guide/migrating-to-2.0.0#language-options

File: docs/rules/no-useless-constructor.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Disallow unnecessary constructor (no-useless-constructor)
+
+ES2015 provides a default constructor if one is not specified. As such, it is unnecessary to provide an empty constructor or one that simply delegates into its parent class.
```
Comment (nzakas) [inline docs/rules/no-useless-constructor.md]: Need more description here. Please show some examples and explain why these constructors are unnecessary.

File: docs/rules/no-useless-constructor.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+## Rule Details
+
+This proposal is to provide a rule to warn against this type of redundant dead code so it can be removed. This rule is similar to `no-empty`.
```
Comment (nzakas) [inline docs/rules/no-useless-constructor.md]: This isn't a proposal, its a rule. 

> This rule flags class constructors that can be safely removed without changing how the class works.

Comment (alberto) [inline docs/rules/no-useless-constructor.md]: Ooops. Fixed.

File: docs/rules/no-useless-constructor.md (hunk 2 above)
Comment (alberto) [inline docs/rules/no-useless-constructor.md]: I added a couple of examples, but I don't know what else to add beyond what's already there.

File: lib/rules/no-useless-constructor.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+    /**
+     * Checks whether the constructor body is a redundant super call.
+     * @param {ASTNode} node - constructor body.
```
Comment (michaelficarra) [inline lib/rules/no-useless-constructor.js]: This doesn't appear to accept a single AST node.

File: lib/rules/no-useless-constructor.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        });
+    }
+
+    /**
+     * Checks whether a node is a redundant construtor
+     * @param {ASTNode} node - node to check
+     * @returns {void}
+     */
+    function checkForConstructor(node) {
+        if (node.key.name !== "constructor") {
```
Comment (michaelficarra) [inline lib/rules/no-useless-constructor.js]: Will this support string keys and not explode when it sees computed keys? I prefer to do type tests when values of multiple types are accepted.

File: lib/rules/no-useless-constructor.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            return;
+        }
+
+        var body = node.value.body.body;
+
+        if (body.length === 0 || isRedundantSuperCall(body, node.value.params)) {
```
Comment (michaelficarra) [inline lib/rules/no-useless-constructor.js]: An empty constructor should only be reported for classes that do not extend anything.

File: lib/rules/no-useless-constructor.js (hunk 5 above)
Comment (alberto) [inline lib/rules/no-useless-constructor.js]: @michaelficarra I see. you mean checking node.kind === "constructor", right?

Comment (michaelficarra) [inline lib/rules/no-useless-constructor.js]: I don't know if that's a thing in the estree AST, but if it is, sure. Just add a test for `class A { "constructor"(){} }`.

**Comment by alberto** [comment]
@michaelficarra Thanks! I updated the rule with your input.

File: lib/rules/no-useless-constructor.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    function checkForConstructor(node) {
+        if (node.kind !== "constructor") {
+            return;
+        }
+
+        var body = node.value.body.body;
+
+        if (!node.parent.parent.superClass && body.length === 0 ||
+            node.parent.parent.superClass && isRedundantSuperCall(body, node.value.params)) {
```
Comment (michaelficarra) [inline lib/rules/no-useless-constructor.js]: `node.parent.parent.superClass ? ... : ...`

**Comment by michaelficarra** [comment]
LGTM.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**